File: genLOS.py

# Functions to generate lines of sight

from mpl_toolkits.mplot3d import Axes3D
import numpy as np
from numpy import linalg as LA
import math
import sys
from random import random, seed
from os import popen
from numpy import linspace as ls
from numba import jit
import subprocess as sp

def inBox(x,y,z,boxsize):
    size = boxsize
    if x<size and x>-1*size and y<size and y>-1*size and z<size and z>-1*size:
        return 1
    else:
        return 0


@jit
def findEnds(px, py, pz, dx, dy, dz, boxsize):
    size = boxsize/2.0
    xen, yen, zen, ten, xex, yex, zex, tex = 0,0,0,0,0,0,0,0
    t = ls(-5e3, 5e3, 5e5)
    
    for i in range(0,len(t)-1):
        x1 = px + dx*t[i]
        y1 = py + dy*t[i]
        z1 = pz + dz*t[i]
        x2 = px + dx*t[i+1]
        y2 = py + dy*t[1+i]
        z2 = pz + dz*t[i+1]
        
        if inBox(x1, y1, z1, size)==0 and inBox(x2, y2, z2, size)==1:
            xen = (x1+x2)/2.0
            yen = (y1+y2)/2.0
            zen = (z1+z2)/2.0
            ten = (t[i]+t[i+1])/2.0
        if inBox(x1, y1, z1, size)==1 and inBox(x2, y2, z2, size)==0:
            xex = (x1+x2)/2.0
            yex = (y1+y2)/2.0
            zex = (z1+z2)/2.0
            tex = (t[i]+t[i+1])/2.0
    return xen, yen, zen, ten, xex, yex, zex, tex

def func(x, y):
    return 0.0

def read_control_file(filename):

    f = open(filename)
    
    for i in range(0,5):
        line = f.readline()

    gasfile = line.split()[0]
    gasfile = gasfile.replace('"', '')
    galID = gasfile.split('_')[0]
    
    line = f.readline()
    rootname = line.split()[0]
    
    line = f.readline()
    aexpn = float(line.split()[0])
    
    for i in range(0,3):
        line = f.readline()
    summaryLoc = (line.split()[0]).replace('"','')

    f.close()

    return gasfile, galID, rootname, aexpn, summaryLoc


def read_summary(galID, aexpn, summaryLoc):

    import sys

    aexpn = float(aexpn)

    # Location of summaries
    f = open(summaryLoc + galID + '.dat')

    f.readline()
    f.readline()

    found = 0

    for line in f:

        l = line.split()
        a = float(l[0])

        if a==aexpn:
            
            found = 1

            # Read in data
            mvir = float(l[2])
            rvir = float(l[3])
            a11 = float(l[4])
            a12 = float(l[5])
            a13 = float(l[6])
            a21 = float(l[7])
            a22 = float(l[8])
            a23 = float(l[9])
            a31 = float(l[10])
            a32 = float(l[11])
            a33 = float(l[12])
            vpec_x = float(l[13])
            vpec_y = float(l[14])
            vpec_z = float(l[15])
            
    f.close()
    if found==1:
        return mvir, rvir, a11, a12, a13, a21, a22, a23, a31, a32, a33, vpec_x, vpec_y, vpec_z
    else:
        print 'ERROR: Could not find the expansion factor {0} in file: \n\t'.format(aexpn) + summaryLoc
        sys.exit()






def genLines(galID, gasfile, summaryLoc, expn, inc, nLOS, maximpact, ncores):

    tol = 1e-5
    seed(25525)
   
    # Read in galaxy properties
    mvir, rvir, a11, a12, a13, a21, a22, a23, a31, a32, a33, vpec_x, vpec_y, vpec_z = read_summary(galID, expn, summaryLoc)
    maximpact_kpc = maximpact * rvir
    
    # Initialize variables
    Xcom, Ycom, Zcom = 0., 0., 0.
    VXcom, VYcom, VZcom = 0., 0., 0.
    l, b = 0.0, 90.0
    vx_obs, vy_obs, vz_obs = 0., 0., 0.
     
    boxsize = 4.0*rvir

    # Open output files
    try:
        fout = open('lines.dat','w')
        finfo = open('lines.info','w')
    except IOError:
        print 'Error opening files lines.dat, lines.info for writing.'
        print 'Check permissions'
        print 'Exitting...'
        sys.exit()

    # Write headers
    fout.write('#        Enter points                      Exit Points\n')
    fout.write('# xen            yen         zen         xex         yex         zex\n')
    finfo.write('# More details on each LOS\n')
    finfo.write('# LOS num     b(kpc)      phi     Inclination\n')

    # Define the rotation matrix to convert from the box frame to the galaxy frame
    a_btg = np.matrix([[a11, a12, a13], [a21, a22, a23], [a31, a32, a33]])

    dxs = []
    dys = []
    dzs = []
    xens = []
    yens = []
    zens = []
    xexs = []
    yexs = []
    zexs = []


    # Generate random impact parameters:
    impacts = []
    for j in range(0,nLOS-1):
        b = maximpact_kpc * random()
        impacts.append(b)

    # Sort impact parameters
    impacts.sort()



    for j in range(0,nLOS):


        b = impacts[j]
        phi = random()*2*math.pi
        
        finfo.write('{0:>6}       {1:>6,.1f}    {2:>6,.1f}   {3:>6,.1f}\n'.format(j+1, b, math.degrees(phi), math.degrees(inc)))
           
        # Define the LOS directional vector in the sky frame:
        ds = np.matrix([0.0, 0.0, -1.0]).T
        
        # Define the impact point in the sky frame:
        ps = np.matrix([b*math.cos(phi), b*math.sin(phi), 0.0]).T


        # Define the rotation matrix to convert from the sky frame to the galaxy frame
        a_stg = np.matrix([[1,0,0], [0, math.cos(inc), -1*math.sin(inc)], [0, math.sin(inc), math.cos(inc)]])

        # Rotate the sky's directional vector and impact point into the galaxy's frame
        dg = a_stg*ds
        pg = a_stg*ps


        # To rotate from the sky's frame to the box frame, need a_gtb, which is the 
        # inverse of a_btg
        a_gtb = LA.inv(a_btg) 
        
        # Rotate the galaxy's directional vector and impact point into the box's frame
        db = a_gtb*dg
        pb = a_gtb*pg


        # Get the enter and exit points of the LOS
        xen, yen, zen, ten, xex, yex, zex, tex = findEnds(pb[0,0], pb[1,0], pb[2,0], db[0,0], db[1,0], db[2,0], boxsize)
        

        fout.write('{0:>12,.5f}{1:>12,.5f}{2:>12,.5f}{3:>12,.5f}{4:>12,.5f}{5:>12,.5f}\n'.format(xen,yen,zen,xex,yex,zex))

        xens.append(xen)
        yens.append(yen)
        zens.append(zen)
        xexs.append(xex)
        yexs.append(yex)
        zexs.append(zex)

    # Determine the inclination of the galaxy
    # Use the dot product of the normal vector and the LOS
    norm_g = np.matrix([0,0,10])
    norm_b = a_gtb*norm_g.T
    los_len = math.sqrt(db[0,0]*db[0,0] + db[1,0]*db[1,0] + db[2,0]*db[2,0])
    norm_len = math.sqrt(norm_b[0,0]*norm_b[0,0] + norm_b[1,0]*norm_b[1,0] + norm_b[2,0]*norm_b[2,0])
    dotprod = db[0,0]*norm_b[0,0] + db[1,0]*norm_b[1,0] + db[2,0]*norm_b[2,0]
    interior = dotprod / (los_len*norm_len)

    # The following conditional statements are needed to account 
    # for rounding errors that occur
    if interior<-1.0 and interior>-1.0-tol:
        interior=-1.0
    elif interior>1.0 and interior<1.0+tol:
        interior=1.0
    angle = math.acos(interior)
    if angle>math.pi/2.0 and angle<math.pi:
        angle = angle - math.pi/2.0

    print 'Calculated inclination: {0:.3f} degrees'.format(math.degrees(angle))

        
    fout.close()
    finfo.close()
   


 
def runCellfinder(codeLoc):
    command = codeLoc+'/funcs/cellfinder/cellfinder'
    try: 
        sp.check_call(command, shell=True)
    except:
        print '\n\nCould not run cellfinder with:\n\t{0:s}'.format(command)
        print 'Exiting...'
        sys.exit()

File: funcs/generateLines/genLOS.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include "matrix.h"
#include "box.h"

int main (int argc, char *argv[]){

    printf("Starting genLOS\n");
    int i, nLOS, ncores;
    double incDeg, maximpact, a; 
    
    // Read in command line arguments
    // 1 = galID
    // 2 = current location where code is being run
    // 3 = expansion parameter
    // 4 = inclination 
    // 5 = number of LOS
    // 6 = maximum impact paramter (in Rvir)
    // 7 = number of cores to use (not used anywhere here)
//    char *galID = argv[1];
    char *summaryLoc = argv[2];
    char *expn = argv[3];
    sscanf(expn, "%lf", &a);
    sscanf(argv[4], "%lf", &incDeg);
    sscanf(argv[5], "%d", &nLOS);
    sscanf(argv[6], "%lf", &maximpact);
    sscanf(argv[7], "%d", &ncores);

    // Convert inclination to radians
    double inc = incDeg * (M_PI/180.0);  

    int tmax = 1000;
    double b, phi, phiDeg;
    double aexpn, redshift;
    double mvir, rvir;
    double a11, a12, a13;
    double a21, a22, a23;
    double a31, a32, a33;
    double vpec_x, vpec_y, vpec_z;
    double galInc;

    double xen, yen, zen, ten;
    double xex, yex, zex, tex;
    double *xens, *yens, *zens;
    double *xexs, *yexs, *zexs;
    double db[3], pb[3], dg[3], pg[3];

    double *ds = (double *)calloc(3, sizeof(double));
    double *ps = (double *)calloc(3, sizeof(double));

    xens = (double *)calloc(nLOS, sizeof(double));
    yens = (double *)calloc(nLOS, sizeof(double));
    zens = (double *)calloc(nLOS, sizeof(double));

    xexs = (double *)calloc(nLOS, sizeof(double));
    yexs = (double *)calloc(nLOS, sizeof(double));
    zexs = (double *)calloc(nLOS, sizeof(double));

    char sumfile[300];
    strcpy(sumfile, summaryLoc);
    strcat(sumfile, "/rotmat_a");
    strcat(sumfile, expn);
    strcat(sumfile, ".txt");
    char new_line[1000];    
    FILE *fp = fopen(sumfile, "r");

    printf("summaryLoc: %s\n",summaryLoc);
    printf("expng: %s\n",expn);
    printf("sumfile: %s\n",sumfile);

    if (fp==NULL){
        printf("\nTest ERROR in genLOS\n");
        printf("Could not open %s\n", sumfile);
        printf("Exitting....\n\n");
        exit(1);
    }
    // Read in the summary file, skipping past the one header row
    fgets(new_line, sizeof(new_line), fp);
    int found = 0;
    while(fgets(new_line, sizeof(new_line), fp)){
        sscanf(new_line, "%lf %lf %lf %lf %lf "
                     "%lf %lf %lf %lf %lf "
                     "%lf %lf %lf %lf %lf "
                     "%lf", 
                      &aexpn, &redshift, &mvir, &rvir, &a11, 
                      &a12, &a13, &a21, &a22, &a23, 
                      &a31, &a32, &a33, &vpec_x, &vpec_y,
                      &vpec_z);
        if ( equals(aexpn,a)){
            found = 1;
        }
    }
    if (found==0){
        printf("\nERROR in genLOS\n");
        printf("Could not find %s in %s\n", expn, sumfile);
        printf("Exitting...\n\n");
        exit(1);
    }
    
    fclose(fp);

    

    // Seed random
    //int seed = 25525;
    //srand(seed);
    srand(time(NULL));

    double maximpact_kpc = maximpact*rvir;
    double boxsize = 4.0*rvir;

    // Open file operators
    FILE *fpdat = fopen("lines.dat", "w");
    FILE *fpinfo = fopen("lines.info", "w");
    
    // Write header
    //fprintf(fpdat, "#        Enter points               Exit Points\n");
    fprintf(fpdat, "# xen      yen         zen      xex      yex       zex\n");
    //fprintf(fpinfo, "# More details on each LOS\n");
    fprintf(fpinfo, "# LOS num        b(kpc)       phi      Inclination\n");



    // Rotation matrix to go from box to galaxy
    double **a_btg, **a_gtb, **a_stg;
    a_btg = (double **)calloc(3, sizeof(double *));
    a_gtb = (double **)calloc(3, sizeof(double *));
    a_stg = (double **)calloc(3, sizeof(double *));
    for (i=0; i<3; i++){
        a_btg[i] = (double *)calloc(3, sizeof(double));
        a_gtb[i] = (double *)calloc(3, sizeof(double));
        a_stg[i] = (double *)calloc(3, sizeof(double));
    }
        
    // Fill a_btg, the rotation matrix for translating from
    // the box frame to the galaxy frome. This is the rotation
    // matrix output by ANA and stored in the summary file
    a_btg[0][0] = a11;
    a_btg[0][1] = a12;
    a_btg[0][2] = a13;
    
    a_btg[1][0] = a21;
    a_btg[1][1] = a22;
    a_btg[1][2] = a23;

    a_btg[2][0] = a31;
    a_btg[2][1] = a32;
    a_btg[2][2] = a33;
    
    // Get the galaxy to box rotation matrix, which is the
    // inversion of the box to galaxy rotation matrix
    inverse(a_btg, a_gtb);

    // Fill ds, the directional vector for the LOS in the sky fram
    ds[0] = 0.0;
    ds[1] = 0.0;
    ds[2] = -1.0;

    // Fill a_stg, the rotation matrix to go from sky to galaxy
    // Depends only on inclination
    a_stg[0][0] = 1.0;    
    a_stg[0][1] = 0.0;    
    a_stg[0][2] = 0.0;    

    a_stg[1][0] = 0.0;    
    a_stg[1][1] = cos(inc);    
    a_stg[1][2] = -1.0*sin(inc);    

    a_stg[2][0] = 0.0;    
    a_stg[2][1] = sin(inc);    
    a_stg[2][2] = cos(inc);    

    // Find dg, the direction vecotr for the LOS in the galaxy frame
    // This is a_stg*ds
    multiply(a_stg, ds, dg);

    // Generate impact parameters
    double impacts[nLOS];
    for(i=0; i<nLOS; i++){
        b = ((double)rand()/(double)RAND_MAX)*maximpact_kpc;
        impacts[i] = b;
    }
    // Sort the impact paramters
    sort_LOS(impacts, nLOS);

    // Loop over LOS
    for(i=0; i<nLOS; i++){

        b = impacts[i];
        phi = ((double)rand()/(double)RAND_MAX)*(2.0*M_PI);
        phiDeg = phi * (180.0/M_PI);
        fprintf(fpinfo, "%d \t %lf \t %lf \t %lf\n",i+1,b,phiDeg,incDeg);

        // Fill ps, the position vector in the sky frame
        ps[0] = b*cos(phi);
        ps[1] = b*sin(phi);
        ps[2] = 0.0;

        // Translate the position vector to the galaxy frame
        // pg = a_stg*ps
        multiply(a_stg, ps, pg);
        
        // Rotate the galaxy's directional vector and impact point 
        // into the box's frame
        // db = a_gtb*dg
        // pb = a_gtb*pg
        multiply(a_gtb, dg, db);
        multiply(a_gtb, pg, pb);
        
        // Find the enter and exit point of the LOS 
        find_ends(pb[0], pb[1], pb[2], db[0], db[1], db[2], boxsize,
                  &xen, &yen, &zen, &xex, &yex, &zex, &ten, &tex, tmax);

        
        if ((xex==0.) || (yex==0.) || (zex==0.) || 
           (xen==0.) || (yen==0.) || (zen==0.)){
            tmax *= 20;
            find_ends(pb[0], pb[1], pb[2], db[0], db[1], db[2], boxsize,
                      &xen, &yen, &zen, &xex, &yex, &zex, &ten, &tex, tmax);
        }


        xens[i] = xen;
        yens[i] = yen;
        zens[i] = zen;
        
        xexs[i] = xex;
        yexs[i] = yex;
        zexs[i] = zex;

        fprintf(fpdat, "%lf \t%lf \t%lf \t%lf \t%lf \t%lf\n",xen,yen,zen,xex,yex,zex);
//        printf("%lf \t%lf \t%lf \t%lf \t%lf \t%lf\n",xen,yen,zen,xex,yex,zex);
//        printf("\n");
    }

    // Test to ensure the inclination of the galaxy is actually what 
    // inc is set to
    galInc = inclination(a_gtb, db);     //radians
    // Report the inclination to screen
//    printf("Calculated inclination angle: %lf degrees\n", galInc*(180.0/M_PI));


    FILE *fptest = fopen("testing.dat", "a");
    fprintf(fptest, "%lf \t %lf\n", inc, galInc);
    fclose(fptest);

    // Close all files
    fclose(fpdat);
    fclose(fpinfo);

    return 0;
}







